Fixed infinite loop FileStorage iter in python 3

Neurostack opened the pull request:
In python 3 if you would iterate over FileStorage and the underlying stream returned bytes arrays the existing sentinel, being the empty string, would never equate to an empty bytes array and an infinite loop would occur. To fix this, I query if the underlying stream inherits from TextIOBase and set the sentinel to an empty string otherwise I set it to an empty bytes array so StopIteration will be called at the end of the file correctly. This is backwards compatible with python 2 as '' == b'' and it does not matter which of these the sentinel is set to.

untitaker commented:
Hey @Neurostack, could you add a test that shows this issue in Py3? Thanks!

untitaker commented:
Also this fixes https://github.com/mitsuhiko/werkzeug/issues/849

Neurostack commented:
@untitaker I think the added test should take care of it.

untitaker commented on tests/test_datastructures.py:
```diff
@@ -900,3 +900,12 @@ class TestFileStorage(object):
+    def test_bytes_proper_sentinel(self):
+        import io
+        unicode_storage = self.storage_class(io.StringIO("one\ntwo"))
+        for idx,line in enumerate(unicode_storage):
```
Bad formatting here.

untitaker commented:
@Neurostack The testcase fails under Python 2, see https://travis-ci.org/pallets/werkzeug/jobs/119346920

Neurostack commented:
@untitaker I installed a version of py2 and fixed StringIO creation and formatting in the test code. I also updated the fix with a more elegant solution suggested by josh ringer in #849. This solution should be more robust as well. It looks like all CI tests now pass.

untitaker commented:
Thanks, please [rebase](https://help.github.com/articles/about-git-rebase/) into one commit.

untitaker commented:
@homu r+

homu commented:
:pushpin: Commit 767485d has been approved by `untitaker`

<!-- @homu r=untitaker 767485dfbcf93c069225e842ace382f314e64c9b -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/pallets/werkzeug/builds/119601852)
